File: python/2020/day_15.py
from pathlib import Path


def play(text: str, target: int) -> int:
    mem = [None] * target
    for turn, n in enumerate(map(int, text.strip().split(',')), start=1):
        mem[n] = turn
    turn, n = len(text.split(',')) + 1, 0
    for turn in range(turn, target):
        diff = turn - mem[n] if mem[n] else 0
        mem[n] = turn
        n = diff
    return n


def main():
    text = (Path(__file__).parent / "../../input/2020/input_15.txt").read_text()
    print("Part 1:", play(text, 2020))
    print("Part 2:", play(text, 30_000_000))


if __name__ == "__main__":
    main()

File: rust/2020/src/bin/day_15.rs
use std::fs;

fn play(nums: &[usize], target: usize) -> usize {
    let mut mem = vec![0; target];
    for (turn, &n) in (1..).zip(nums) {
        mem[n] = turn;
    }
    let mut n = 0;
    for turn in (nums.len() + 1)..target {
        let diff = match mem[n] {
            0 => 0,
            t => turn - t,
        };
        mem[n] = turn;
        n = diff;
    }
    n
}

fn main() {
    let input = fs::read_to_string("../input/2020/input_15.txt").unwrap();
    let nums: Vec<_> = input
        .trim()
        .split(',')
        .map(|s| s.parse().unwrap())
        .collect();
    println!("Part 1: {}", play(&nums, 2020));
    println!("Part 2: {}", play(&nums, 30_000_000));
}
